torch/onnx/_internal/exporter.py

```diff
@@ -0,0 +1,248 @@
+        self.model = model
+        self.model_args = model_args
+        self.model_kwargs = model_kwargs
+
+    @abc.abstractmethod
+    def run(self) -> ExportOutput:
+        pass
+
+    @property
+    def logger(self) -> logging.Logger:
+        # options.logger will always be resolved to an instance when constructing
+        assert isinstance(self.options.logger, logging.Logger)
+        return self.options.logger
+
+
+@_beartype.beartype
+def _resolve_export_options(options: Optional[ExportOptions]) -> ExportOptions:
+    if options is None:
+        options = ExportOptions()
+
+    T = TypeVar("T")
+
+    def resolve(value: Optional[T], fallback: Callable[[], T]):
+        return fallback() if value is None else value
+
+    return ExportOptions(
+        opset_version=resolve(options.opset_version, lambda: _DEFAULT_OPSET_VERSION),
+        dynamic_shapes=options.dynamic_shapes,
+        logger=resolve(
+            options.logger, lambda: logging.getLogger().getChild("torch.onnx")
+        ),
+    )
+
+
+@_beartype.beartype
+def dynamo_export(
+    model: torch.nn.Module,
+    /,
+    *model_args,
+    export_options: Optional[ExportOptions] = None,
+    **model_kwargs,
+) -> ExportOutput:
+    """Export a torch.nn.Module to an ONNX graph.
+
+    Args:
+        model: The PyTorch model to be exported to ONNX.
+        model_args: Positional inputs to ``model``.
+        model_kwargs: Keyword inputs to ``model``.
+        export_options: Options to influence the export to ONNX.
+
+    Returns:
+        An in-memory representation of the exported ONNX model.
+
+    Example:
+    ::
+
+        import torch.onnx
+        torch.onnx.dynamo_export(
+            my_nn_module,
+            torch.randn(2, 2, 2), # positional input 1
+            torch.randn(2, 2, 2), # positional input 2
+            my_nn_module_attribute="hello", # keword input
+            export_options=ExportOptions(
+                opset_version=17,
+            )
+        ).save("my_model.onnx")
+    """
+
+    resolved_export_options = _resolve_export_options(export_options)
+
+    def assert_dependencies():
+        # this has been resolved above, so will not be 'None'
+        assert (opset_version := resolved_export_options.opset_version) is not None
+
+        # TODO: each of these imports/assertions should be turned into
+        # an actionable diagnostic, e.g.:
+        # - Please `pip install onnxscript_torchlib`
+        # - You have specified an opset version X that is newer than the
+        #   supported opset version Y
+        # - etc.
+        import onnx
+
+        # mypy is unhappy with 'import onnxscript' right now:
+        # - https://github.com/microsoft/onnx-script/issues/572
+        import onnxscript  # type: ignore[import]
+
+        assert onnx.defs.onnx_opset_version() >= opset_version
+        assert isinstance(
+            onnxscript.onnx_opset.all_opsets[("", opset_version)],
+            onnxscript.values.Opset,
+        )
```

Comment:
```suggestion
        ), "The installed ONNX Script does not support the specified opset version. Install a newer ONNX Script or specify an older opset version!"
```

Reply:
This can be part of the PR that plumb through the FX exporter. Adding implementation and hardening checks.

* Raise `ValueError`
* Raise with formatted error message with user specified version, and acceptable range.
* Update docstring with `Raises:`

Reply:
I am inclined to agree, but this PR is already adding some implementation. Just maikng sure that whatever is added, is added in the best possible way (giving the partial support for v1)

Reply:
Updated.